src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/CheckstyleJavaParser.g4

```diff
@@ -0,0 +1,858 @@
+      DOT LITERAL_CLASS                                                    #primitivePrimary
+    | nonWildcardTypeArguments LITERAL_THIS arguments                      #paramPrimary
+    ;
+
+primaryClassOrInterfaceTypeNoArray
+    : annotations[false] classOrInterfaceType[false]
+    ;
+
+primaryPrimitiveTypeNoArray
+    : annotations[false] primitiveType
+    ;
+
+classType
+    : (classOrInterfaceType[false] DOT)? annotations[false] id typeArguments?
+    ;
+
+creator
+    : nonWildcardTypeArguments createdName classCreatorRest
+    | annotations[false] createdName
+      (annotations[false] arrayCreatorRest | classCreatorRest)
+    ;
+
+createdName
+    : id typeArgumentsOrDiamond? extended+=createdNameExtended*            #createdNameObject
+    | primitiveType                                                        #createdNamePrimitive
+    ;
+
+createdNameExtended
+    : DOT annotations[false] id typeArgumentsOrDiamond?
+    ;
+
+innerCreator
+    : annotations[false] id nonWildcardTypeArgumentsOrDiamond? classCreatorRest
+    ;
+
+arrayCreatorRest
+    : LBRACK
+      ( RBRACK arrayDeclarator* arrayInitializer
+      | expression RBRACK bracketsWithExp* arrayDeclarator*
+      )
+    ;
+
+bracketsWithExp
+    : annotations[false] LBRACK expression RBRACK
+    ;
+
+classCreatorRest
+    : arguments classBody?
+    ;
+
+typeArgumentsOrDiamond
+    : LT GT                                                                #diamond
+    | typeArguments                                                        #typeArgs
+    ;
+
+nonWildcardTypeArgumentsOrDiamond
+    : LT GT                                                                #nonWildcardDiamond
+    | nonWildcardTypeArguments                                             #nonWildcardTypeArgs
+    ;
+
+nonWildcardTypeArguments
+    : LT typeArgumentsTypeList GT
+    ;
+
+typeArgumentsTypeList
+    : typeType[false] (COMMA typeType[false])*
+    ;
+
+typeList
+    : typeType[false] (COMMA typeType[false])*
+    ;
+
+typeType[boolean createImaginaryNode]
+    : annotations[false] classOrInterfaceOrPrimitiveType arrayDeclarator*
+    ;
+
+classOrInterfaceOrPrimitiveType
+    : (classOrInterfaceType[false] | primitiveType)
+    ;
+
+arrayDeclarator
+    : anno=annotations[false] LBRACK RBRACK
+    ;
+
+primitiveType
+    : LITERAL_BOOLEAN
+    | LITERAL_CHAR
+    | LITERAL_BYTE
+    | LITERAL_SHORT
+    | LITERAL_INT
+    | LITERAL_LONG
+    | LITERAL_FLOAT
+    | LITERAL_DOUBLE
+    | LITERAL_VOID
+    ;
+
+typeArguments
+    : LT typeArgument (COMMA typeArgument)* GT
+    ;
+
+superSuffix
+    : LPAREN expressionList? RPAREN                                        #superSuffixSimple
+    | DOT id (LPAREN expressionList? RPAREN)?                              #superSuffixDot
+    ;
+
+arguments
+    : LPAREN expressionList? RPAREN
+    ;
+
+patternDefinition
+    : patternVariableDefinition
+    ;
+
+patternVariableDefinition
+    : mods+=modifier* type=typeType[true] id
```

Comment:
The ANTLR4 `+=` operator creates a list of rule contexts. This makes for easy iteration, and reduces the need for null checks because we just return an empty collection if there are no elements that match.